File: java/uk/ac/mrc/hgu/Wlz/Test/WlzBoundingBox.java
import java.io.*;
import uk.ac.mrc.hgu.Wlz.*;

public class WlzBoundingBox
{
  public static void main (String[] args)
  {
    boolean	showUsageFlg = false,
    		verboseFlg = false;
    int		optIdx = 0,
    		exitStatus = 0;
    String outFile = "-";

    while((optIdx < args.length) && (args[optIdx].startsWith("-")) &&
          (args[optIdx].length() > 1))
    {
      if(args[optIdx].equals("-h"))
      {
        showUsageFlg = true;
      }
      else if(args[optIdx].equals("-o"))
      {
        if((optIdx + 1) < args.length)
	{
	  outFile = args[optIdx + 1];
	  ++optIdx;
	  if(outFile.length() < 1)
	  {
	    showUsageFlg = true;
	    exitStatus = 1;
	  }
	}
      }
      else if(args[optIdx].equals("-v"))
      {
        verboseFlg = true;
      }
      else
      {
        showUsageFlg = true;
	exitStatus = 1;
      }
      ++optIdx;
    }
    if(showUsageFlg)
    {
      System.err.println("Usage: WlzBoundingBox [-o <out file>] [-v] [-h] " +
      			 "[<input files>]");
      System.err.println(
      "Gets the background value of the given Woolz object(s).");
      System.err.println("Example: WlzBoundingBox -o out.txt myobj.wlz");
      System.err.println(
      "The input Woolz object is read from myobj.wlz. The bounding box");
      System.err.println("is calculated and  written to out.txt.");
      System.exit(exitStatus);
    }
    else
    {
      try
      {
	boolean	firstPass = true;
	String inFile = "-";
	if(outFile.equals("-") == false)
	{
	  System.setOut(new PrintStream(
	  		new FileOutputStream(outFile)));
	}
	while(firstPass || (optIdx < args.length))
	{
          WlzFileStream in = null;
	  if((firstPass && (optIdx == args.length)) ||
	     args[optIdx].equals("-"))
          {
	    inFile = "-";
	    in = new WlzFileStdInStream();
	  }
	  else
	  {
	    inFile = args[optIdx];
	    in = new WlzFileInputStream(args[optIdx]);
	  }
          WlzObject obj0 = WlzObject.WlzReadObj(in);
	  WlzIBox3 bBox = WlzObject.WlzBoundingBox3D(obj0);
	  if(verboseFlg)
	  {
	    System.out.println("file: " + inFile);
	    System.out.println("xMin: " + bBox.xMin);
	    System.out.println("yMin: " + bBox.yMin);
	    System.out.println("zMin: " + bBox.zMin);
	    System.out.println("xMax: " + bBox.xMax);
	    System.out.println("yMax: " + bBox.yMax);
	    System.out.println("zMax: " + bBox.zMax);
	  }
	  else
	  {
	    System.out.println(bBox.xMin + " " +
	    		       bBox.yMin + " " +
	    		       bBox.zMin + " " +
	    		       bBox.xMax + " " +
	    		       bBox.yMax + " " +
	    		       bBox.zMax);
	  }
	  firstPass = false;
	  ++optIdx;
	}
      }
      catch (IOException e)
      {
	System.err.println(e);
	System.exit(1);
      }
      catch (WlzException e)
      {
	System.err.println(e);
	System.exit(1);
      }
    }
  }
}

File: binWlz/WlzBoundingBox.c
#pragma ident "MRC HGU $Id$"
/*!
* \file		binWlz/WlzBoundingBox.c
* \author       Bill Hill
* \date         March 1999
* \version      $Id$
* \par
* Address:
*               MRC Human Genetics Unit,
*               Western General Hospital,
*               Edinburgh, EH4 2XU, UK.
* \par
* Copyright (C) 2005 Medical research Council, UK.
* 
* This program is free software; you can redistribute it and/or
* modify it under the terms of the GNU General Public License
* as published by the Free Software Foundation; either version 2
* of the License, or (at your option) any later version.
*
* This program is distributed in the hope that it will be
* useful but WITHOUT ANY WARRANTY; without even the implied
* warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR
* PURPOSE.  See the GNU General Public License for more
* details.
*
* You should have received a copy of the GNU General Public
* License along with this program; if not, write to the Free
* Software Foundation, Inc., 51 Franklin Street, Fifth Floor,
* Boston, MA  02110-1301, USA.
* \brief	calculates the bounding box of an object.
* \ingroup	BinWlz
* \todo         -
* \bug          None known.
*
* \par Binary
* \ref wlzboundingbox "WlzBoundingBox"
*/

/*!
\ingroup      BinWlz
\defgroup     wlzboundingbox WlzBoundingBox
\par Name
WlzBoundingBox - Calculates the bounding box of the given object.
\par Synopsis
\verbatim
WlzBoundingBox [-o <output file>] [-h] [<input object file>]

\endverbatim
\par Options
<table width="500" border="0">
  <tr>
    <td><b>-o</b></td>
    <td>output file name</td>
  </tr>
  <tr>
    <td><b>-h</b></td>
    <td>Help - print help message</td>
  </tr>
</table>
By default the input object is read from the standard input
and the output is written to the standard output.

\par Description
WlzBoundingBox calculates the bounding box
of the given Woolz object.
The bounding box is written in the following order:
\<x min\> \<y min\> \<z min\> \<x max\> \<y max\> \<z max\>

\par Examples
\verbatim
WlzBoundingBox reads an object from myobj.wlz, calculates
its bounding box and then prints it to the standard output.

WlzBoundingBox myobj.wlz
\endverbatim
\par File
\ref WlzBoundingBox.c "WlzBoundingBox.c"
\par See Also
\ref BinWlz "WlzIntro(1)"
\ref wlzfacts "WlzFacts(1)"
\ref WlzBoundingBox2I "WlzBoundingBox2I(3)"
\ref WlzBoundingBox2D "WlzBoundingBox2D(3)"
\ref WlzBoundingBox3I "WlzBoundingBox3I(3)"
\ref WlzBoundingBox3D "WlzBoundingBox3D(3)"
*/

#ifndef DOXYGEN_SHOULD_SKIP_THIS
#include <stdio.h>
#include <stdlib.h>
#include <ctype.h>
#include <string.h>
#include <Wlz.h>

extern int      getopt(int argc, char * const *argv, const char *optstring);
 
extern char     *optarg;
extern int      optind,
                opterr,
                optopt;

int             main(int argc, char **argv)
{
  int		option,
		ok = 1,
		usage = 0;
  WlzErrorNum	errNum = WLZ_ERR_NONE;
  FILE		*fP = NULL;
  WlzObject	*inObj = NULL;
  WlzDBox3	bBox;
  char 		*outFileStr,
  		*inObjFileStr;
  static char	optList[] = "o:h",
		outFileStrDef[] = "-",
  		inObjFileStrDef[] = "-";

  opterr = 0;
  outFileStr = outFileStrDef;
  inObjFileStr = inObjFileStrDef;
  while(ok && ((option = getopt(argc, argv, optList)) != -1))
  {
    switch(option)
    {
      case 'o':
        outFileStr = optarg;
	break;
      case 'h':
      default:
        usage = 1;
	ok = 0;
	break;
    }
  }
  if((inObjFileStr == NULL) || (*inObjFileStr == '\0') ||
     (outFileStr == NULL) || (*outFileStr == '\0'))
  {
    ok = 0;
    usage = 1;
  }
  if(ok && (optind < argc))
  {
    if((optind + 1) != argc)
    {
      usage = 1;
      ok = 0;
    }
    else
    {
      inObjFileStr = *(argv + optind);
    }
  }
  if(ok)
  {
    if((inObjFileStr == NULL) ||
       (*inObjFileStr == '\0') ||
       ((fP = (strcmp(inObjFileStr, "-")?
	      fopen(inObjFileStr, "r"): stdin)) == NULL) ||
       ((inObj= WlzReadObj(fP, &errNum)) == NULL) ||
       (errNum != WLZ_ERR_NONE))
    {
      ok = 0;
      (void )fprintf(stderr,
		     "%s: failed to read object from file %s\n",
		     *argv, inObjFileStr);
    }
    if(fP && strcmp(inObjFileStr, "-"))
    {
      fclose(fP);
    }
  }
  if(ok)
  {
    bBox = WlzBoundingBox3D(inObj, &errNum);
    if(errNum != WLZ_ERR_NONE)
    {
      ok = 0;
      (void )fprintf(stderr,
      		     "%s: failed to compute bounding box of object\n",
		     *argv);
    }
  }
  if(ok)
  {
    if(((fP = (strcmp(outFileStr, "-")?
              fopen(outFileStr, "w"):
	      stdout)) == NULL))
    {
      ok = 0;
      (void )fprintf(stderr,
      		     "%s: failed to write output data\n",
		     *argv);
    }
    else
    {
      (void )fprintf(fP, "%g %g %g %g %g %g\n",
      		     bBox.xMin, bBox.yMin, bBox.zMin,
		     bBox.xMax, bBox.yMax, bBox.zMax);
    }
    if(fP && strcmp(outFileStr, "-"))
    {
      fclose(fP);
    }
  }
  if(usage)
  {
    (void )fprintf(stderr,
    "Usage: %s%sExample: %s%s",
    *argv,
    " [-o<out file>] [-h] [<in object>]\n"
    "Options:\n"
    "  -o  Output file name.\n"
    "  -h  Help, prints this usage message.\n"
    "Calculates the bounding box of the input Woolz object. The data\n"
    "are written to the output file in the following order:\n"
    "  <x min> <y min> <z min> <x max> <y max> <z max>\n"
    "The input object is read from stdin and data are written to stdout\n"
    "unless the filenames are given.\n",
    *argv,
    " -o out.txt myobj.wlz\n"
    "The input Woolz object is read from myobj.wlz. It's bounding box\n"
    "is computed and written to out.txt.\n");
  }
  return(!ok);
}
#endif /* DOXYGEN_SHOULD_SKIP_THIS */
